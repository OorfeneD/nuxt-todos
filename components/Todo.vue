<template>
  <li :class="`border border-gray-100 text-gray-500 rounded-md my-1 pb-2 ${todo.status ? 'bg-gray-100' : ''}`">
    <div class="pt-2 px-3 flex">
      <input
        :id="todo.id"
        class="flex-grow-0 cursor-pointer"
        :checked="todo.status"
        type="checkbox"
        name="todo"
        @change="$emit('updatetodo', { todo, status: $event.target.checked, originalEvent: $event})"
      >
      <label class="flex-grow text-center pl-1 pr-3 cursor-pointer" :for="todo.id">
        <strike v-if="todo.status" class="text-gray-500 transition transition-all">
          {{ todo.content.title }}
        </strike>
        <span v-else>
          {{ todo.content.title }}
        </span>
      </label>
      <button class="font-bold text-red-800 flex-grow-0 -my-2 -mx-3 px-3 hover:bg-red-400 hover:text-white transition rounded-r" @click="$emit('removetodo', { todo: todo, originalEvent: $event })">
        <icon close />
      </button>
    </div>
    <div v-if="!!todo.tags" class="flex flex-row px-1 pt-1">
      <tag-list class="pl-2" :tags="tags.filter(v => todo.tags.includes(v.id))" />
    </div>
  </li>
</template>

<script>
import TagList from './tags/TagList.vue'
export default {
  components: { TagList },
  props: {
    todo: {
      type: Object,
      default: () => ({
        id: -1,
        content: {
          title: '<i>dead inside</i>'
        }
      })
    },
    tags: {
      type: Array,
      default: () => []
    }
  }
}
</script>

<style>

</style>
