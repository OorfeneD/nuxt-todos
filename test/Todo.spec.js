import { shallowMount } from '@vue/test-utils'
import Todo from '../components/Todo.vue'
import { titleText, todo as todoPropData } from './mocks/todo.mock'
describe('Todo Component', () => {
  const wrapper = shallowMount(Todo, {
    propsData: {
      todo: todoPropData
    }
  })
  const emptyWrapper = shallowMount(Todo)
  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })
  test('is renders with no props', () => {
    expect(emptyWrapper.vm).toBeTruthy()
    expect(emptyWrapper.vm.todo).not.toBeUndefined()
  })
  test('instance contains `todo` prop data', () => {
    expect(wrapper.vm).toStrictEqual(expect.objectContaining({ todo: todoPropData }))
  })
  test('instance renders title', () => {
    expect(wrapper.text()).toContain(titleText)
  })
  test('matchs snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  test('emits event on checkbox state changing', () => {
    wrapper.find('input').trigger('click')
    wrapper.find('input').trigger('change')
    expect(wrapper.emitted().updatetodo).toBeTruthy()
  })
  test('status wasn`t changed after checkbox click', () => {
    expect(wrapper.vm.todo.status).not.toBeTruthy()
  })
  test('emitted event contains `status` string with `true` value', () => {
    expect(wrapper.emitted('updatetodo')[0]).not.toBeUndefined()
    expect(wrapper.emitted('updatetodo')[0][0]).toStrictEqual(
      expect.objectContaining(
        { todo: todoPropData, newStatus: true }
      )
    )
  })
  test('emits event on checkbox state changing', () => {
    wrapper.find('button').trigger('click')
    expect(wrapper.emitted().removetodo).toBeTruthy()
  })
})
