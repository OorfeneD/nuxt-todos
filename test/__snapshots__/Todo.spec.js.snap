// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Component matchs snapshot 1`] = `
"<li class=\\"py-2 px-3 border border-gray-100 text-gray-500 rounded-md my-1 flex  \\"><input id=\\"hash\\" type=\\"checkbox\\" name=\\"todo\\" class=\\"flex-grow-0\\"> <label for=\\"hash\\" class=\\"flex-grow text-center pl-1 pr-3\\"><span>
      todo text
    </span></label> <button class=\\"font-bold text-red-800 flex-grow-0 -my-2 -mx-3 px-3 border-l hover:bg-red-400 hover:text-white transition rounded-r\\">
    <icon-stub close=\\"true\\"></icon-stub>
  </button></li>"
`;
